### ISSUE TITLE
listen tcp : protocol not available on Raspbian

### ISSUE THREAD

**Issue by missdeer** [coredns/coredns#2239]
Hi,
I have tried to run CoreDNS on Raspbian, Corefile is:

```
.:54545 {
  forward . 8.8.8.8
  log
  health
}
```
Then it complains `2018/10/25 21:21:47 listen tcp :54545: protocol not available`, changing the port number doesn't help.

**Comment by chrisohaver** [comment]
Can you listen on ports with `nc` e.g. `nc -l 54545` ?

**Comment by missdeer** [comment]
@chrisohaver Yes, nc is ok.

**Comment by miekg** [comment]
Could this be the so-reuse port bit, somehow?

**Comment by missdeer** [comment]
@miekg 
I'm not sure. How to check it?
BTW, other programs such as dnsmasq can listen on the same port for both TCP and UDP protocols there, and works fine.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] listen tcp : ..." ]
>@miekg
>I'm not sure. How to check it?
>BTW, other programs such as dnsmasq can listen on the same port for both TCP and UDP protocols there, and works fine.

~~~
HypriotOS/armv7: pirate@node0 in /tmp
$ ./coredns  -dns.port=1053
.:1053
2018/10/26 06:17:47 [INFO] CoreDNS-1.2.5
2018/10/26 06:17:47 [INFO] linux/arm, go1.11.1, 204537b
CoreDNS-1.2.5
linux/arm, go1.11.1, 204537b
~~~

And: `dig @node0 -p 1053 a whoami.miek.nl`

~~~
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;whoami.miek.nl.			IN	A

;; ADDITIONAL SECTION:
whoami.miek.nl.		0	IN	A	192.168.87.28
_udp.whoami.miek.nl.	0	IN	SRV	0 0 36865 .
~~~

**Comment by missdeer** [comment]
@miekg 
```
pi@bananapi-r1:~/coredns_1.2.5 $ ./coredns -dns.port=1053
2018/10/26 14:31:15 listen tcp :1053: protocol not available
```

**Comment by miekg** [comment]
Then I don't know


/label: needs info

**Comment by missdeer** [comment]
Is it possible to listen on UDP port only?

**Comment by miekg** [comment]
No, that's not supported.

**Comment by stuartnelson3** [comment]
Check your kernel, `uname -r`. `SO_REUSEPORT` was first included in kernel version 3.9:

https://lwn.net/Articles/542629/

The latest code in the dns library assumes that if the code is compiled with go1.11, then it should use `SO_REUSEPORT`, which does not take into consideration that the feature is actually dependent on the underlying kernel supporting it.

I've encountered this problem on an old machine running a 3.2 kernel.

**Comment by chrisohaver** [comment]
Thanks @stuartnelson3, good to know!

**Comment by missdeer** [comment]
Seemly I have met the same issue:

```
pi@bananapi-r1:~ $ uname -r
3.4.112-sun7i
```

So the only solution is upgrading my kernel?

**Comment by stuartnelson3** [comment]
You can:

- Upgrade your kernel
- Compile coredns with go1.10
- Use an older version of coredns and wait for the underlying dns library to be patched

**Comment by miekg** [comment]
can you try https://github.com/coredns/coredns/pull/2244 ?

**Comment by missdeer** [comment]
@miekg 
Patched #2244 , it could run now, but dig failed:

```
pi@bananapi-r1:~/coredns_1.2.5 $ ./coredns -dns.port=1053
2018/10/28 07:05:50 [WARNING] Failed to set SO_REUSEPORT on socket: protocol not available
2018/10/28 07:05:50 [WARNING] Failed to set SO_REUSEPORT on socket: protocol not available
127.0.0.1.:1053
2018/10/28 07:05:50 [INFO] CoreDNS-1.2.5
2018/10/28 07:05:50 [INFO] linux/arm, go1.11.1, 
CoreDNS-1.2.5
linux/arm, go1.11.1, 
```

```
pi@bananapi-r1:~/coredns_1.2.5 $ netstat -anlp | grep coredns
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
tcp6       0      0 :::1053                 :::*                    LISTEN      14051/coredns   
tcp6       0      0 :::8080                 :::*                    LISTEN      14051/coredns   
udp6       0      0 :::1053                 :::*                                14051/coredns   
pi@bananapi-r1:~/coredns_1.2.5 $ dig @127.0.0.1 -p 1053 www.google.com

; <<>> DiG 9.9.5-9+deb8u16-Raspbian <<>> @127.0.0.1 -p 1053 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 44357
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;www.google.com.			IN	A

;; Query time: 3 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sun Oct 28 07:10:56 CST 2018
;; MSG SIZE  rcvd: 43

pi@bananapi-r1:~/coredns_1.2.5 $ dig @::1 -p 1053 www.google.com

; <<>> DiG 9.9.5-9+deb8u16-Raspbian <<>> @::1 -p 1053 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 40333
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;www.google.com.			IN	A

;; Query time: 3 msec
;; SERVER: ::1#1053(::1)
;; WHEN: Sun Oct 28 07:11:06 CST 2018
;; MSG SIZE  rcvd: 43

pi@bananapi-r1:~/coredns_1.2.5 $ dig @::1 -p 1053 www.twitter.com

; <<>> DiG 9.9.5-9+deb8u16-Raspbian <<>> @::1 -p 1053 www.twitter.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 11791
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;www.twitter.com.		IN	A

;; Query time: 3 msec
;; SERVER: ::1#1053(::1)
;; WHEN: Sun Oct 28 07:11:41 CST 2018
;; MSG SIZE  rcvd: 44

pi@bananapi-r1:~/coredns_1.2.5 $ dig @127.0.0.1 -p 1053 www.twitter.com

; <<>> DiG 9.9.5-9+deb8u16-Raspbian <<>> @127.0.0.1 -p 1053 www.twitter.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 18227
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;www.twitter.com.		IN	A

;; Query time: 3 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sun Oct 28 07:11:50 CST 2018
;; MSG SIZE  rcvd: 44

pi@bananapi-r1:~/coredns_1.2.5 $ dig @127.0.0.1 -p 1053 www.baidu.com

; <<>> DiG 9.9.5-9+deb8u16-Raspbian <<>> @127.0.0.1 -p 1053 www.baidu.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 26684
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;www.baidu.com.			IN	A

;; Query time: 3 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sun Oct 28 07:11:57 CST 2018
;; MSG SIZE  rcvd: 42

pi@bananapi-r1:~/coredns_1.2.5 $ dig @::1 -p 1053 www.baidu.com

; <<>> DiG 9.9.5-9+deb8u16-Raspbian <<>> @::1 -p 1053 www.baidu.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 33787
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;www.baidu.com.			IN	A

;; Query time: 3 msec
;; SERVER: ::1#1053(::1)
;; WHEN: Sun Oct 28 07:12:06 CST 2018
;; MSG SIZE  rcvd: 42

```

**Comment by missdeer** [comment]
@miekg 
Sorry, please ignore my previous comment, #2244 works fine, dig failure was caused by my incorrect Corefile.

@stuartnelson3 
BTW, using go 1.10.4 without #2244 could resolve this issue too.

Thanks everyone.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] listen tcp : ..." ]
>@miekg
>Sorry, please ignore my previous comment, #2244 works fine, dig failure was caused by my incorrect Corefile.
>
>@stuartnelson3
>BTW, using go 1.10.4 without #2244 could resolve this issue too.
>
>Thanks everyone.

Thanks for the ping back, something ala #2244 will go in.

**Comment by miekg** [comment]
Fixed via #2243
